winpr/libwinpr/clipboard/synthetic_file.c

```diff
@@ -307,91 +286,73 @@ static BOOL add_directory_entry_to_list(wClipboard* clipboard, const char* local
 	return result;
 }
 
-static BOOL do_add_directory_contents_to_list(wClipboard* clipboard, const char* local_name,
-                                              const WCHAR* remote_name, DIR* dirp,
+static BOOL do_add_directory_contents_to_list(wClipboard* clipboard, const WCHAR* local_name,
+                                              const WCHAR* remote_name, HANDLE hFind,
                                               wArrayList* files)
 {
-	/*
-	 * For some reason POSIX does not require readdir() to be thread-safe.
-	 * However, readdir_r() has really insane interface and is pretty bad
-	 * replacement for it. Fortunately, most C libraries guarantee thread-
-	 * safety of readdir() when it is used for distinct directory streams.
-	 *
-	 * Thus we can use readdir() in multithreaded applications if we are
-	 * sure that it will not corrupt some global data. It would be nice
-	 * to have a compile-time check for this here, but some C libraries
-	 * do not provide a #define because of reasons (I'm looking at you,
-	 * musl). We should not be breaking people's builds because of that,
-	 * so we do nothing and proceed with fingers crossed.
-	 */
-	for (;;)
+	if (INVALID_HANDLE_VALUE == hFind)
 	{
-		struct dirent* entry = NULL;
-		errno = 0;
-		entry = readdir(dirp);
+		return FALSE;
+	}
 
-		if (!entry)
+	while (TRUE)
+	{
+		WIN32_FIND_DATAW FileData;
+		BOOL bRet = FindNextFileW(hFind, &FileData);
+		if (!bRet)
 		{
-			int err = errno;
-
-			if (!err)
-				break;
-
-			WLog_ERR(TAG, "failed to read directory: %s", strerror(err));
+			WLog_WARN(TAG, "FindNextFile failed (%d)", GetLastError());
+			if (ERROR_NO_MORE_FILES == GetLastError())
+				return TRUE;
 			return FALSE;
 		}
 
-		if (!add_directory_entry_to_list(clipboard, local_name, remote_name, entry, files))
+		if (!add_directory_entry_to_list(clipboard, local_name, remote_name, &FileData, files))
 			return FALSE;
 	}
 
 	return TRUE;
 }
 
-static BOOL add_directory_contents_to_list(wClipboard* clipboard, const char* local_name,
+static BOOL add_directory_contents_to_list(wClipboard* clipboard, const WCHAR* local_name,
                                            const WCHAR* remote_name, wArrayList* files)
 {
 	BOOL result = FALSE;
-	DIR* dirp = NULL;
-	WLog_VRB(TAG, "adding directory: %s", local_name);
-	dirp = opendir(local_name);
+	WCHAR namebuf[MAX_PATH];
+	HANDLE hFind = NULL;
+	WIN32_FIND_DATAW FindData;
 
-	if (!dirp)
+	swprintf(namebuf, L"%s/*.*", local_name);
```

Comment:
`Be careful with `L'somestring'`, windows understands `UCS-2/UTF16` most other systems `UCS4/UTF32` so the resulting string will be broken on many systems